src/networking.c

```diff
@@ -1401,12 +1402,26 @@ void acceptCommonHandler(connection *conn, struct ClientFlags flags, char *ip) {
         return;
     }
 
+    if (connAddrSockName(conn, NULL, 0, &port)) {
+        serverLog(LL_WARNING, "Unable to retrieve socket port");
+        connClose(conn);
+        return;
+    }
+
+    if (port == server.admin_port && connIsLocal(conn) != 1) {
+        serverLog(LL_WARNING, "Denied connection. On admin-port, connections are"
+                              " only accepted from the loopback interface.");
+        server.stat_rejected_conn++;
+        connClose(conn);
+        return;
+    }
+
     /* Limit the number of connections we take at the same time.
      *
      * Admission control will happen before a client is created and connAccept()
      * called, because we don't want to even start transport-level negotiation
      * if rejected. */
-    if (listLength(server.clients) + getClusterConnectionsCount() >= server.maxclients) {
+    if (port != server.admin_port && (listLength(server.clients) + getClusterConnectionsCount() >= server.maxclients)) {
```

Comment:
connections from `admin-port` can consume clients out of `maxclients`, but the eventloop's size is fixed, do we need to make eventloop to be auto resizing?

Reply:
We can add a fixed number and limit the number of admin connections. We don't need to count the admin connections separately. We can just accept connections on the admin port as long as the total number of connections is less than for example `server.maxclients + 20`.

Reply:
Added a check for the number of connections accepted on the port. Currently, it's set to 20, but I believe 10 would be sufficient. Additionally, the event loop size also been updated based on the maximum number of clients accepted on the port.

Reply:
@PingXie Why is that a problem? Do you think clients are surprised they can make fewer connections because an admin is connected?

Or do you think the admins need to be completely invisible to other clients, like not shown in CLIENT LIST, etc.?

I think the admins can be visible and I don't see why it's a problem that a admin eating the users' quota is a problem.

Reply:
At amazon we have a fixed number of clients ontop of what is requested as well (although I believe we chose 256). It's worth mentioning that clusterbus connections also eat into the overhead.

Reply:
I think this feature could solve part of this problem in cluster mode.

Reply:
> At amazon we have a fixed number of clients ontop of what is requested as well (although I believe we chose 256). It's worth mentioning that clusterbus connections also eat into the overhead.

We also have a similar CONFIG_EXTRA_RESERVED_FDS, the value is 500. we are doing some whitelist-ips(and localhost) to bypass the maxclients check. I am worry that admin-port consume a listen port and in here it only used for bypass the maxclients check. I am hoping there is a way that admin can do more, for example, when the server is blocked, admin can still has the way to connect in.

Reply:
> I am hoping there is a way that admin can do more,

Yeah we discussed in yesterday's core team meeting that the fd limit needs to be taken care as well. In addition, we should not evict "admin clients" either. I also think we shouldn't allow a "normal" or "standard" client to kill "admin" clients. Another related question would be whether we should count the "admin" clients when reporting "connected_clients" in `INFO`.

For the fd limit, I think we could avoid hard-code values by resizing the `eventloop->events` array dynamically in `aeCreateFileEvent` when the current size is about to be exceeded. The main benefit though would be simplified error handling - see #528

Reply:
> In addition, we should not evict "admin clients" either. I also think we shouldn't allow a "normal" or "standard" client to kill "admin" clients. Another related question would be whether we should count the "admin" clients when reporting "connected_clients" in INFO.

this feels a bit overkill, i agree this is better. A admin should be very fast, and the number should be quite small. Normally, we should try to avoid using the admin port. Ps, we marked client kill as a admin command, so a normal user is unlikely to call it to kill the admin client (we can add a type filter for sure)

> For the fd limit, I think we could avoid hard-code values by resizing the eventloop->events array dynamically in aeCreateFileEvent when the current size is about to be exceeded. 

yes, i agree dynamically is better in this case.